provide a working example for memoize

wilzbach opened the pull request:
Turns out that arrays are objects and for some reasons are different - is this a bug in the memoization?

Anyways I found a lot better example that is easier to understand, _works_ and shows off another great feature of D.
However for the debug statement to be printed it depends on #181, it needs

```
rdmd -debug
```

Ping @ZombineDev

Btw I couldn't find a nice & simple way to print the BigInt in scientific notation - ideas?

ZombineDev commented:
Instead of (ab)using debug, you can just showcase the main reason to use memoize in first place - to save execution time. Here's an alternative example:

```
import std.bigint;

/**
Computes the power of a base with an exponent.

Returns:
    Result of the power as an
    arbitrary-sized integer
*/
BigInt bigpow(uint base, uint exp) pure
{
    // to help debugging pure functions
    // can contain impure debug code

    BigInt result = base;
    while (exp > 0)
    {
        result *= base;
        exp--;
    }
    return result;
}

void main()
{
    import std.stdio;
    import std.functional: memoize;
    import std.conv : text;
    import std.datetime: benchmark, to;

    // memoize caches the result of the function
    // call depending on the input parameters.
    // pure functions are great for that!
    alias fastBigPow = memoize!(bigpow);

    void test()
    {
        auto len = fastBigPow(5, 10000).text.length;
        writef("calculating length = %s took: ", len);
    }

    foreach (i; 0 .. 10)
        benchmark!test(1)[0].to!("msecs", float).writeln;
}
```

stonemaster commented:
I prefer @ZombineDev's solution because the result is at least as convincing that `memoize` works and shows the real-world benefit of how D can makes things perform better, still just requiring one a few bytes more. I also ran it on the tour online and performance gains are visible there too.

wilzbach commented:
> Instead of (ab)using debug, you can just showcase the main reason to use memoize in first place - to save execution time. Here's an alternative example:

That's a great idea!
I rebased with this and made two minor modifications
- documentation header was now greater than 48 chars -> line break
- I moved "took" to the benchmark function to avoid confusion

ZombineDev commented:
> I moved "took" to the benchmark function to avoid confusion

That was also how I did it at first, but it made `benchmark` a couple of magnitudes slower, which made the `memoize` win by a much smaller margin - something like 20ms vs 1-2ms (instead of 0.005ms).

wilzbach commented:
> That was also how I did it at first, but it made benchmark a couple of magnitudes slower, which made the memoize win by a much smaller margin - something like 20ms vs 1-2ms (instead of 0.005ms).

Sorry for the confusion - I just moved the console output printing word "took" to the `foreach` loop. At least on my machine this doesn't affect the huge gained speed-up.

stonemaster commented on public/content/en/gems.md:
```diff
@@ -450,31 +450,45 @@ and `@nogc`).
-    // result will be cached!
-    writeln(fastMedian([7, 5, 3]));
+    foreach (i; 0 .. 10)
+        benchmark!test(1)[0].to!("msecs", double).writeln("took: miliseconds");
```
48 chars :-)

wilzbach commented:
> 48 chars :-)

There were actually more occurences.
Fixed them and filled #196.
Btw we also should fix the mobile view (#44) to actually get advantage of the 48 chars rule.
